AnnoDesigner.Core/DataStructures/QuadTree.cs

```diff
@@ -357,154 +359,184 @@ public void ReIndex(Func<T, Rect> boundsSelector)
         public QuadTree(Rect extent)
         {
             root = new Quadrant(extent);
-            Extent = extent;
         }
 
         /// <summary>
         /// Insert a <typeparamref name="T"/> item into the QuadTree.
         /// </summary>
-        /// <param name="item"></param>
-        /// <param name="bounds"></param>
-        public void Insert(T item, Rect bounds)
+        public void Add(T item)
         {
-            if (!root.Extent.Contains(bounds))
-            {
-                throw new ArgumentException($"{nameof(bounds)} of {nameof(item)} is greater than the extent of the quadtree.");
-            }
-            root.Insert(item, bounds);
+            var bounds = item.Bounds;
+            EnsureBounds(bounds);
+
+            root.Add(item, bounds);
         }
 
         /// <summary>
         /// Remove a <typeparamref name="T"/> item from the QuadTree.
         /// </summary>
-        /// <param name="item"></param>
-        /// <param name="bounds"></param>
-        public void Remove(T item, Rect bounds)
+        public bool Remove(T item)
         {
-            var quadrant = root.GetContainingQuadrant(bounds);
-            quadrant.Remove((item, bounds));
+            return root.Remove(item);
         }
 
         /// <summary>
-        /// Retrieves a sequence of items that intersect the given bounds.
+        /// Removes all items from the QuadTree.
         /// </summary>
-        /// <param name="bounds">The bounds to check</param>
-        /// <returns></returns>
-        public IEnumerable<T> GetItemsIntersecting(Rect bounds)
+        public void Clear()
         {
-            var items = new List<T>(previousCount);
-            root.GetItemsIntersecting(items, bounds);
-            previousCount = items.Count;
-            return items;
+            foreach (var item in this.ToList())
```

Comment:
Shouldn't this also be lazily evaluated? Or is this `ToList()`ed explicitly for a specific reason? If it does need to be materialised, then its worth adding a comment as to why.

Reply:
It was there because otherwise the iterated collection would change, invalidating iterators. But it was mainly written like this to enable registering undoable operations with the event system and since that is not on the table anymore, I'll modify this to just create new Extent.